src/c++/perf_analyzer/docs/examples/profile.py

```diff
@@ -175,73 +223,134 @@ def collect_periodic_latencies(args, filename):
     return bins
 
 
-def calculate_avg_periodic_latencies(args, filename):
+def calculate_avg_periodic_latencies(args, profile_result, filename):
     """Calculate average token-to-token latency for each request period."""
     bins = collect_periodic_latencies(args, filename)
 
     latencies = []
     for bin in bins:
-        latencies.append(np.mean(bin) / 1_000_000_000)
-    return latencies
+        latencies.append(np.mean(bin) / 1_000_000)
+
+    profile_result.avg_periodic_t2t_latencies = latencies
 
 
 def collect_latencies(requests):
     # Example json demonstrating format:
     #   see client/src/c++/perf_analyzer/docs/examples/decoupled_output_file.json
     first_token_latencies = []
+    generation_latencies = []
     token_to_token_latencies = []
     requests = requests["experiments"][0]["requests"]
+    for r in requests:
+        init_request, responses = r["timestamp"], r["response_timestamps"]
+        first_token_latencies.append((responses[0] - init_request) / 1_000_000)
+        generation_latencies.append((responses[-1] - responses[0]) / 1_000_000)
+        token_to_token_latencies = []
+        for prev_res, res in pairwise(responses):
+            token_to_token_latencies.append((res - prev_res) / 1_000_000)
+    return first_token_latencies, generation_latencies, token_to_token_latencies
+
+
+def calculate_online_metrics(args, profile_result, filename):
+    """Calculate online metrics for more fine-grained performance information."""
+    if not args.stream:
+        return  # skip if offline
+
+    requests = load_json_data(filename)
```

Comment:
Is the same file ever loaded more than once? Can we avoid that by reading it once and passing the file pointer/dict or whatever down to these functions?

Reply:
Good point. Changed the code to read data only once.